src/messaging/src/index.ts

```diff
@@ -0,0 +1,537 @@
+    // Handle the common case of a non-conflatable message.
+    if (!msg.isConflatable) {
+      enqueueMessage(handler, msg);
+      return;
+    }
+
+    // Conflate the message with an existing message if possible.
+    let conflated = some(messageQueue, posted => {
+      if (posted.handler !== handler) {
+        return false;
+      }
+      if (!posted.msg) {
+        return false;
+      }
+      if (posted.msg.type !== msg.type) {
+        return false;
+      }
+      if (!posted.msg.isConflatable) {
+        return false;
+      }
+      return posted.msg.conflate(msg);
+    });
+
+    // Enqueue the message if it was not conflated.
+    if (!conflated) {
+      enqueueMessage(handler, msg);
+    }
+  }
+
+  /**
+   * Install a message hook for a message handler.
+   *
+   * @param handler - The message handler of interest.
+   *
+   * @param hook - The message hook to install.
+   *
+   * #### Notes
+   * A message hook is invoked before a message is delivered to the
+   * handler. If the hook returns `false`, no other hooks will be
+   * invoked and the message will not be delivered to the handler.
+   *
+   * The most recently installed message hook is executed first.
+   *
+   * If the hook is already installed, this is a no-op.
+   */
+  export
+  function installMessageHook(handler: IMessageHandler, hook: MessageHook): void {
+    // Lookup the hooks for the handler.
+    let hooks = messageHooks.get(handler);
+
+    // Bail early if the hook is already installed.
+    if (hooks && hooks.indexOf(hook) !== -1) {
+      return;
+    }
+
+    // Add the hook to the end, so it will be the first to execute.
+    if (!hooks) {
+      messageHooks.set(handler, [hook]);
+    } else {
+      hooks.push(hook);
+    }
+  }
+
+  /**
+   * Remove an installed message hook for a message handler.
+   *
+   * @param handler - The message handler of interest.
+   *
+   * @param hook - The message hook to remove.
+   *
+   * #### Notes
+   * It is safe to call this function while the hook is executing.
+   *
+   * If the hook is not installed, this is a no-op.
+   */
+  export
+  function removeMessageHook(handler: IMessageHandler, hook: MessageHook): void {
+    // Lookup the hooks for the handler.
+    let hooks = messageHooks.get(handler);
+
+    // Bail early if the hooks do not exist.
+    if (!hooks) {
+      return;
+    }
+
+    // Lookup the index of the hook and bail if not found.
+    let i = hooks.indexOf(hook);
+    if (i === -1) {
+      return;
+    }
+
+    // Clear the hook and schedule a cleanup of the array.
+    hooks[i] = null;
+    scheduleCleanup(hooks);
+  }
+
+  /**
+   * Clear all message data associated with a message handler.
+   *
+   * @param handler - The message handler of interest.
+   *
+   * #### Notes
+   * This will clear all posted messages and hooks for the handler.
+   */
+  export
+  function clearData(handler: IMessageHandler): void {
```

Comment:
They used to be called `clearMessageData`, `clearSignalData`, and `clearPropertyData` because they were non-namespaced. The longer form would seem redundant to me now that they are namespaced.